src/frontend/src/scheduler/hummock_snapshot_manager.rs

```diff
@@ -144,49 +155,37 @@ struct HummockSnapshotManagerCore {
     epoch_to_query_ids: BTreeMap<u64, HashSet<QueryId>>,
     meta_client: Arc<dyn FrontendMetaClient>,
     last_unpin_snapshot: Arc<AtomicU64>,
+    max_committed_epoch: Arc<AtomicU64>,
 }
 
 impl HummockSnapshotManagerCore {
-    fn new(meta_client: Arc<dyn FrontendMetaClient>) -> Self {
+    fn new(meta_client: Arc<dyn FrontendMetaClient>, max_committed_epoch: Arc<AtomicU64>) -> Self {
         Self {
             // Initialize by setting `is_outdated` to `true`.
             meta_client,
             epoch_to_query_ids: BTreeMap::default(),
             last_unpin_snapshot: Arc::new(AtomicU64::new(0)),
+            max_committed_epoch,
         }
     }
 
-    async fn get_epoch_for_query(
+    fn get_epoch_for_query(
         &mut self,
         batches: &mut Vec<(QueryId, Callback<SchedulerResult<u64>>)>,
     ) -> u64 {
-        let ret = self.meta_client.get_epoch().await;
-        match ret {
-            Ok(epoch) => {
-                let queries = match self.epoch_to_query_ids.get_mut(&epoch) {
-                    None => {
-                        self.epoch_to_query_ids.insert(epoch, HashSet::default());
-                        self.epoch_to_query_ids.get_mut(&epoch).unwrap()
-                    }
-                    Some(queries) => queries,
-                };
-                for (id, cb) in batches.drain(..) {
-                    queries.insert(id);
-                    let _ = cb.send(Ok(epoch));
-                }
-                epoch
-            }
-            Err(e) => {
-                for (id, cb) in batches.drain(..) {
-                    let _ = cb.send(Err(SchedulerError::Internal(anyhow!(
-                        "Failed to get epoch for query: {:?} because of RPC Error: {:?}",
-                        id,
-                        e
-                    ))));
-                }
-                0
+        let epoch = self.max_committed_epoch.load(Ordering::Relaxed);
```

Comment:
Although we will synchronously fetch epoch on observer manager startup, I am still a bit concerned that future changes may miss to update max_committed_epoch before using it in the query, which will cause incorrect result. Can we add a debug assert here to make sure `epoch != INVALID_EPOCH`?

Reply:
Seems we cannot add this assertion, because if I just run one query `select 1 as one` the epoch will be 0, causing frontend to crash on debug mode 🤣